File: Assemblies/Protean.Tools/Counter.vb
' ================================================================================================
'   Protean.Tools.Counter & Protean.Tools.CounterCollection
'   Desc:       A simple counter class as well as a CounterCollection
'   Author:     Ali Granger
'   Updated:    03-Feb-10
'   Docs:       
'
'   
'
' ================================================================================================
Imports System
Imports System.Collections


Public Class Counter

#Region "Declarations"

    Private _name As String
    Private _count As Double
    Private _base As Double
    Private _increment As Double

#End Region

#Region "Enums and Constants"

    Public Const DEFAULT_INCREMENT As Double = 1
    Public Const DEFAULT_BASE As Double = 0

#End Region

#Region "Events"

    Public Event OnChange(ByVal sender As Counter, ByVal e As EventArgs)

    Public Shadows Event OnError(ByVal sender As Object, ByVal e As Protean.Tools.Errors.ErrorEventArgs)

    Private Sub _OnError(ByVal sender As Object, ByVal e As Protean.Tools.Errors.ErrorEventArgs)
        RaiseEvent OnError(sender, e)
    End Sub

#End Region

#Region "Constructor"

    Public Sub New()
        Me.New(Tools.Text.RandomPassword(16, , TextOptions.LowerCase Or TextOptions.UseAlpha Or TextOptions.UseNumeric), DEFAULT_INCREMENT, DEFAULT_BASE)
    End Sub

    Public Sub New(ByVal name As String)
        Me.New(name, DEFAULT_INCREMENT, DEFAULT_BASE)
    End Sub

    Public Sub New(ByVal name As String, ByVal incrementInterval As Integer)
        Me.New(name, Convert.ToDouble(incrementInterval), DEFAULT_BASE)
    End Sub

    Public Sub New(ByVal name As String, ByVal incrementInterval As Double)
        Me.New(name, incrementInterval, DEFAULT_BASE)

    End Sub

    Public Sub New(ByVal name As String, ByVal incrementInterval As Double, ByVal baseCounter As Double)
        _name = name
        Increment = incrementInterval
        _base = baseCounter
        Reset()
    End Sub


#End Region

#Region "Private Properties"


#End Region

#Region "Public Properties"
    Public ReadOnly Property Name() As String
        Get
            Return _name
        End Get
    End Property

    Public ReadOnly Property Count() As Double
        Get
            Return _count
        End Get
    End Property


    Public ReadOnly Property Base() As Double
        Get
            Return _base
        End Get
    End Property


    Public Property Increment() As Double
        Get
            Return _increment
        End Get
        Set(ByVal value As Double)
            _increment = value
        End Set
    End Property

#End Region

#Region "Private Members"


#End Region

#Region "Public Members"

    Public Sub Reset()
        Reset(DEFAULT_BASE)
    End Sub

    Public Sub Reset(ByVal baseCounter As Integer)
        Convert.ToDouble(baseCounter)
    End Sub

    Public Sub Reset(ByVal baseCounter As Double)
        _count = baseCounter
        RaiseEvent OnChange(Me, Nothing)
    End Sub

    Public Sub Add()
        Add(Increment)
    End Sub

    Public Sub Add(ByVal incremementInterval As Integer)
        Add(Convert.ToDouble(incremementInterval))
    End Sub

    Public Sub Add(ByVal incremementInterval As Double)
        _count += incremementInterval
        RaiseEvent OnChange(Me, Nothing)
    End Sub

    Public Function ToInt() As Integer

        Return Convert.ToInt32(_count)

    End Function

#End Region

#Region "Shared Members"


#End Region

End Class ' Counter

Public Class CounterCollection
    Inherits CollectionBase

#Region "Inherited Overrides"

    Default Public Property Item(ByVal index As Integer) As Counter
        Get
            Return CType(List(index), Counter)
        End Get
        Set(ByVal value As Counter)
            List(index) = value
        End Set
    End Property

    Default Public Property Item(ByVal counterName As String) As Counter
        Get
            Return CType(List(IndexOf(counterName)), Counter)
        End Get
        Set(ByVal value As Counter)
            List(IndexOf(counterName)) = value
        End Set
    End Property

    ''' <summary>
    ''' Creates a counter by counter name and adds it to the collection
    ''' </summary>
    ''' <param name="counterName">The name of the counter</param>
    ''' <returns>Returns the actual Counter object</returns>
    ''' <remarks></remarks>
    Public Function Add(ByVal counterName As String) As Counter
        If Not Exists(counterName) Then
            Return Item(List.Add(New Counter(counterName)))
        Else
            Throw New Exception("Counter name already exists")
            Return Nothing
        End If
    End Function 'Add

    ''' <summary>
    ''' Adds a counter by object
    ''' </summary>
    ''' <param name="value">The counter object to add</param>
    ''' <returns>The index of counter in the collection</returns>
    ''' <remarks></remarks>
    Public Function Add(ByVal value As Counter) As Integer
        Return List.Add(value)
    End Function 'Add


    Public Function IndexOf(ByVal counterName As String) As Integer
        Dim index As Integer = -1
        For Each counterItem As Counter In List
            If counterItem.Name = counterName Then
                index = IndexOf(counterItem)
            End If
        Next
        Return index
    End Function 'IndexOf

    Public Function IndexOf(ByVal value As Counter) As Integer
        Return List.IndexOf(value)
    End Function 'IndexOf


    Public Sub Insert(ByVal index As Integer, ByVal value As Counter)
        List.Insert(index, value)
    End Sub 'Insert


    Public Sub Remove(ByVal value As Counter)
        List.Remove(value)
    End Sub 'Remove


    Public Function Contains(ByVal value As Counter) As Boolean
        ' If value is not of type Counter, this will return false.
        Return List.Contains(value)
    End Function 'Contains


    Protected Overrides Sub OnInsert(ByVal index As Integer, ByVal value As [Object])
        ' Insert additional code to be run only when inserting values.
    End Sub 'OnInsert


    Protected Overrides Sub OnRemove(ByVal index As Integer, ByVal value As [Object])
        ' Insert additional code to be run only when removing values.
    End Sub 'OnRemove


    Protected Overrides Sub OnSet(ByVal index As Integer, ByVal oldValue As [Object], ByVal newValue As [Object])
        ' Insert additional code to be run only when setting values.
    End Sub 'OnSet


    Protected Overrides Sub OnValidate(ByVal value As [Object])
        If Not value.GetType() Is Type.GetType("Protean.Tools.Counter") Then
            Throw New ArgumentException("value must be of type Protean.Tools.Counter.", "value")
        End If
    End Sub 'OnValidate 

#End Region

#Region "Public Non-inherited Members"

    Public Function Exists(ByVal value As String) As Boolean
        Dim isFound As Boolean = False
        For Each counterItem As Counter In List
            If counterItem.Name = value Then
                isFound = True
                Exit For
            End If
        Next
        Return isFound
    End Function 'Exists

    Public Sub ResetAll()
        For Each counterItem As Counter In List
            counterItem.Reset()
        Next
    End Sub

#End Region

End Class 'CounterCollection

File: Assemblies/Protean.Tools.Csharp/Counter.cs
// ================================================================================================
// Protean.Tools.Counter & Protean.Tools.CounterCollection
// Desc:       A simple counter class as well as a CounterCollection
// Author:     Ali Granger
// Updated:    03-Feb-10
// ================================================================================================
using System;
using System.Collections;
using static Protean.Tools.Text;

namespace Protean.Tools
{
    public class Counter
    {

        #region Declarations

        private string _name;
        private double _count;
        private double _base;
        private double _increment;

        #endregion

        #region Enums and Constants

        public const double DEFAULT_INCREMENT = 1d;
        public const double DEFAULT_BASE = 0d;

        #endregion

        #region Events

        public event OnChangeEventHandler OnChange;

        public delegate void OnChangeEventHandler(Counter sender, EventArgs e);

        public new event OnErrorEventHandler OnError;

        public new delegate void OnErrorEventHandler(object sender, Errors.ErrorEventArgs e);

        private void _OnError(object sender, Errors.ErrorEventArgs e)
        {
            OnError?.Invoke(sender, e);
        }

        #endregion

        #region Constructor

        public Counter() : this(RandomPassword(16, options: TextOptions.LowerCase | TextOptions.UseAlpha | TextOptions.UseNumeric), DEFAULT_INCREMENT, DEFAULT_BASE)
        {
        }

        public Counter(string name) : this(name, DEFAULT_INCREMENT, DEFAULT_BASE)
        {
        }

        public Counter(string name, int incrementInterval) : this(name, Convert.ToDouble(incrementInterval), DEFAULT_BASE)
        {
        }

        public Counter(string name, double incrementInterval) : this(name, incrementInterval, DEFAULT_BASE)
        {
        }

        public Counter(string name, double incrementInterval, double baseCounter)
        {
            _name = name;
            Increment = incrementInterval;
            _base = baseCounter;
            Reset();
        }


        #endregion

        #region Private Properties


        #endregion

        #region Public Properties
        public string Name
        {
            get
            {
                return _name;
            }
        }

        public double Count
        {
            get
            {
                return _count;
            }
        }

        public double Base
        {
            get
            {
                return _base;
            }
        }

        public double Increment
        {
            get
            {
                return _increment;
            }

            set
            {
                _increment = value;
            }
        }

        #endregion

        #region Private Members


        #endregion

        #region Public Members

        public void Reset()
        {
            Reset(DEFAULT_BASE);
        }

        public void Reset(int baseCounter)
        {
            Convert.ToDouble(baseCounter);
        }

        public void Reset(double baseCounter)
        {
            _count = baseCounter;
            OnChange?.Invoke(this, null);
        }

        public void Add()
        {
            Add(Increment);
        }

        public void Add(int incremementInterval)
        {
            Add(Convert.ToDouble(incremementInterval));
        }

        public void Add(double incremementInterval)
        {
            _count += incremementInterval;
            OnChange?.Invoke(this, null);
        }

        public int ToInt()
        {
            return Convert.ToInt32(_count);
        }

        #endregion

        #region Shared Members


        #endregion

    } // Counter

    public class CounterCollection : CollectionBase
    {

        #region Inherited Overrides

        public Counter this[int index]
        {
            get
            {
                return (Counter)List[index];
            }

            set
            {
                List[index] = value;
            }
        }

        public Counter this[string counterName]
        {
            get
            {
                return (Counter)List[IndexOf(counterName)];
            }

            set
            {
                List[IndexOf(counterName)] = value;
            }
        }

        /// <summary>
        /// Creates a counter by counter name and adds it to the collection
        /// </summary>
        /// <param name="counterName">The name of the counter</param>
        /// <returns>Returns the actual Counter object</returns>
        /// <remarks></remarks>
        public Counter Add(string counterName)
        {
            if (!Exists(counterName))
            {
                return this[List.Add(new Counter(counterName))];
            }
            else
            {
                throw new Exception("Counter name already exists");
                return null;
            }
        } // Add

        /// <summary>
        /// Adds a counter by object
        /// </summary>
        /// <param name="value">The counter object to add</param>
        /// <returns>The index of counter in the collection</returns>
        /// <remarks></remarks>
        public int Add(Counter value)
        {
            return List.Add(value);
        } // Add

        public int IndexOf(string counterName)
        {
            int index = -1;
            foreach (Counter counterItem in List)
            {
                if ((counterItem.Name ?? "") == (counterName ?? ""))
                {
                    index = IndexOf(counterItem);
                }
            }

            return index;
        } // IndexOf

        public int IndexOf(Counter value)
        {
            return List.IndexOf(value);
        } // IndexOf

        public void Insert(int index, Counter value)
        {
            List.Insert(index, value);
        } // Insert

        public void Remove(Counter value)
        {
            List.Remove(value);
        } // Remove

        public bool Contains(Counter value)
        {
            // If value is not of type Counter, this will return false.
            return List.Contains(value);
        } // Contains

        protected override void OnInsert(int index, object value)
        {
            // Insert additional code to be run only when inserting values.
        } // OnInsert

        protected override void OnRemove(int index, object value)
        {
            // Insert additional code to be run only when removing values.
        } // OnRemove

        protected override void OnSet(int index, object oldValue, object newValue)
        {
            // Insert additional code to be run only when setting values.
        } // OnSet

        protected override void OnValidate(object value)
        {
            if (!ReferenceEquals(value.GetType(), Type.GetType("Protean.Tools.Counter")))
            {
                throw new ArgumentException("value must be of type Protean.Tools.Counter.", "value");
            }
        } // OnValidate 

        #endregion

        #region Public Non-inherited Members

        public bool Exists(string value)
        {
            bool isFound = false;
            foreach (Counter counterItem in List)
            {
                if ((counterItem.Name ?? "") == (value ?? ""))
                {
                    isFound = true;
                    break;
                }
            }

            return isFound;
        } // Exists

        public void ResetAll()
        {
            foreach (Counter counterItem in List)
                counterItem.Reset();
        }

        #endregion

    } // CounterCollection
}